osgbuild/fetch_sources.py

```diff
@@ -32,194 +82,263 @@
+    return pos,neg
+
+def kvmatch(arg):
+    # return (key,val) for "key=val", else return (None, arg)
+    return re.search(r'^(?:(\w+)=)?(.*)', arg).groups()
+
+def fancy_source_error(meta_type, explicit_type, handler, args, kw, e):
+    xtype = "type" if explicit_type else "implicit type"
+    log.error("Error processing source line of %s '%s'" % (xtype, meta_type))
+    varnames = handler.__code__.co_varnames
+    fn_argcount = handler.__code__.co_argcount
+    maxargs = varnames.index('ops')
+    posargs = varnames[:maxargs]
+    posargs_provided = posargs[:len(args)]
+    dupe_args = set(posargs_provided) & set(kw)
+
+    pos_usage = ' '.join("[%s=]arg%s" % (a,i+1) for i,a in enumerate(posargs))
+    log.error("Up to %s unnamed initial arguments are allowed: %s"
+              % (maxargs, pos_usage))
+
+    if dupe_args:
+        for arg in dupe_args:
+            log.error("No unnamed positional arguments allowed after"
+                      " explicit '%s' named field" % arg)
+    elif len(args) > maxargs:
+        log.error("Provided %s positional arguments: %s"
+                  % (len(args), ' '.join(args)))
+    else:
+        log.error(e)
+    raise Error("Invalid parameters for %s=%s source line" % (xtype,meta_type))
+
+def process_dot_source(cache_prefix, sfilename, destdir, nocheck, want_spec):
+    """Read a .source file, fetch any sources specified in it."""
+    ops = FetchOptions(destdir=destdir, cache_prefix=cache_prefix,
+                       nocheck=nocheck, want_spec=want_spec)
 
-    """
     utils.safe_makedirs(destdir)
-    downloaded = []
-    with open(sfilename, 'r') as sfile:
-        for lineno, line in enumerate(sfile):
-            line = line.strip()
-            if line.startswith('#'):
-                continue
-            if line == '':
-                continue
-            if len(line.split()) > 1:
-                filenames = process_meta_url(line, destdir, nocheck)
-                downloaded.extend(filenames)
-                continue
-            elif line.startswith('/'):
-                uri = "file://" + line
-                log.warning(
-                    "An absolute path has been given in %s line %d. "
-                    "It is recommended to use only paths relative to %s"
-                    "in your source files.", sfilename, lineno+1,
-                    cache_prefix)
-            elif not re.match(r'/|\w+://', line): # relative path
-                uri = os.path.join(cache_prefix, line)
-            else:
-                uri = line
-
-            log.info('Retrieving ' + uri)
-            try:
-                handle = urllib.request.urlopen(uri)
-            except urllib.error.URLError as err:
-                raise Error("Unable to download %s\n%s" % (uri, str(err)))
-            filename = os.path.join(destdir, os.path.basename(line))
+    filenames = []
+    for line in open(sfilename):
+        line = re.sub(r'(^|\s)#.*', '', line).strip()
```

Comment:
The only reason I know that this line skips comments and blank lines is because that's what the code you're replacing did. Either add a comment or simplify the code.

Reply:
>skips comments and blank lines
>comment or simplify
Roger!